internal/controllers/fake/upsert.go

```diff
@@ -0,0 +1,63 @@
+package fake
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Helper methods for making updates.
+
+func UpsertSpec(ctx context.Context, t testing.TB, ctrlClient ctrlclient.Client, obj ctrlclient.Object) {
+	t.Helper()
+
+	require.True(t, obj.GetName() != "", "object has empty name")
+	err := ctrlClient.Create(ctx, obj)
+	if ctx.Err() != nil {
+		return
+	}
+	if err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			return
+		}
+		assert.NoError(t, err)
```

Comment:
This reads funny to me -- I know you're asserting no error here because there is an error and you want to raise it. I wonder if this works a bit better.
```suggestion
		assert.Fail(t, "create failed", "%+v %+v", obj, err)
```